modules/beefy/src/lib.rs

```diff
@@ -0,0 +1,855 @@
+		fn on_initialize(_n: T::BlockNumber) -> frame_support::weights::Weight {
+			<RequestCount<T, I>>::mutate(|count| *count = count.saturating_sub(1));
+
+			(0_u64)
+				.saturating_add(T::DbWeight::get().reads(1))
+				.saturating_add(T::DbWeight::get().writes(1))
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config<I>, I: 'static> Pallet<T, I>
+	where
+		BridgedBeefyMmrHasher<T, I>: 'static + Send + Sync,
+	{
+		/// Initialize pallet with BEEFY validator set and best finalized block number.
+		#[pallet::weight((T::DbWeight::get().reads_writes(2, 4), DispatchClass::Operational))]
+		pub fn initialize(
+			origin: OriginFor<T>,
+			init_data: InitializationDataOf<T, I>,
+		) -> DispatchResult {
+			ensure_owner_or_root::<T, I>(origin)?;
+
+			let init_allowed = !<BestBlockNumber<T, I>>::exists();
+			ensure!(init_allowed, <Error<T, I>>::AlreadyInitialized);
+
+			log::info!(target: "runtime::bridge-beefy", "Initializing bridge BEEFY pallet: {:?}", init_data);
+			Ok(initialize::<T, I>(init_data)?.into())
+		}
+
+		/// Halt or resume all pallet operations.
+		///
+		/// May only be called either by root, or by `PalletOwner`.
+		#[pallet::weight((T::DbWeight::get().reads_writes(1, 1), DispatchClass::Operational))]
+		pub fn set_operational(origin: OriginFor<T>, operational: bool) -> DispatchResult {
+			ensure_owner_or_root::<T, I>(origin)?;
+			<IsHalted<T, I>>::put(!operational);
+
+			if operational {
+				log::info!(target: "runtime::bridge-beefy", "Resuming pallet operations");
+			} else {
+				log::warn!(target: "runtime::bridge-beefy", "Stopping pallet operations");
+			}
+
+			Ok(().into())
+		}
+
+		/// Change `PalletOwner`.
+		///
+		/// May only be called either by root, or by `PalletOwner`.
+		#[pallet::weight((T::DbWeight::get().reads_writes(1, 1), DispatchClass::Operational))]
+		pub fn set_owner(
+			origin: OriginFor<T>,
+			new_owner: Option<T::AccountId>,
+		) -> DispatchResultWithPostInfo {
+			ensure_owner_or_root::<T, I>(origin)?;
+			match new_owner {
+				Some(new_owner) => {
+					PalletOwner::<T, I>::put(&new_owner);
+					log::info!(target: "runtime::bridge-beefy", "Setting pallet Owner to: {:?}", new_owner);
+				},
+				None => {
+					PalletOwner::<T, I>::kill();
+					log::info!(target: "runtime::bridge-beefy", "Removed Owner of pallet.");
+				},
+			}
+
+			Ok(().into())
+		}
+
+		/// Submit commitment, generated by BEEFY validators.
+		///
+		/// Apart from the generic payload, the commitment contains the finalized (by BEEFY) block
+		/// number, so they must be always be imported in order. Importing commitment gives us
+		/// knowledge of header hash that has been finalized by BEEFY validators.
+		#[pallet::weight(0)] // TODO: compute weights
+		pub fn submit_commitment(
+			origin: OriginFor<T>,
+			// TODO: implement `TypeInfo` for `BridgedBeefySignedCommitment<T, I>`, `BeefyMmrProof`
+			// and `BridgedBeefyMmrLeafUnpacked::<T, I>`
+			encoded_commitment: Vec<u8>,
```

Comment:
While working on relay, I've found having mandatory `encoded_mmr_proof` and `mmr_leaf` arguments here breaks our basic assumptions: relay is not putting any restrictions on the node it is connected to. But in this case (unless I'm totally wrong), we need to be connected to node that has offchain indexing enabled and to generate a proof we need to have unpruned state (because it is generated during runtime call, iiuc).

The first requirement seems adequate, because it is how BEEFY is designed. The second one is redundant imo. So probably we shall:
1) make these two arguments optional, or even better - extract to a separate call;
2) get rid, or make optional this "unpacked" validator set structure. We shall be using merkle root of validator keys instead and optionally - save proved validator keys to the storage (to avoid verifying same keys next time when the same set is signing something).

Whether this will be fixed in this, or in follow up PRs, depends on how things will be going (review vs relay development).

Reply:
> But in this case, we need to be connected to node that has offchain indexing enabled and to generate a proof we need to have unpruned state.

Correct, given the relayer stays completely stateless. Alternatively the relayer could be fetching and storing this data as it is produced (even without finality) and use that instead of relying on the node to provide it.

Another option to get rid of the archive node requirement is to make the MMR proof creation completely off-chain - I believe the data in off-chain storage should always be sufficient (i.e. it's never pruned), but for simplicity we do that as a runtime call. We could move some of that logic to the relayer itself (sure, "hardcoding" trade-offs again).

The validator set is a bit easier, since it does not change that frequently. Also if the BEEFY digest item stays, we could use validators from there (i.e. not rely on the state) and re-merkelize them (again, "hardcoding" that logic in the relay).

IMHO these proposals are critical to consider long-term, but if delivery time is important it might be better to simply add the extra offchain-indexing + archive requirement to the upstream node, with an intention to optimize this later.


> save proved validator keys to the storage (to avoid verifying same keys next time when the same set is signing something).

Depending on the acceptable bridge latency (aka cost), we might end up importing just one `SignedCommitment` per session so this cache might not be very useful. However it obviously is a valid solution.

Reply:
Thank you! So there are regular (GRANDPA-like) authorities-related digest items in BEEFY - I've missed it somehow (only seen this MMR root digest). So probably it would be better to revert to the same schema we've been using with GRANDPA && to track authorities from digest items. In this case leaf and its proof are definitely optional - they may only be required when we need to verify some secondary proof (like storage proof of some parachain -> parachain heads root -> mmr leaf -> mmr leaf proof).

I've seen @acatangiu was working on removing MMR root digest item. But I've found no issues about authorities-related digests removal (there's one related to BABE validators within parachains, but it seems unrealted), so I'll probably take this option. Thank you!

Reply:
In our sub2sub bridge I think we can rely on the BEEFY authorities-related digest items in headers and avoid relying on the "more complicated" leaf data and its proof. There are no plans to remove the auth-related digest from substrate chains headers; the digest happens just once per session so its overhead is negligible.

Other non-substrate-based-chains bridges will use the BEEFY payload, the leafs and their proofs.

Reply:
My bad - it is impossible to build BEEFY light client using only headers + commitments. There's nothing about header in the commitment (apart from block number, which is useless in this case), so we don't have a link between commitment and header => can't rely on header digests. So MMR leafs + proofs are definitely essential for the client. I'll try to experiment with generating MMR proofs outside of runtime - it should solve the archive-node problem.

Reply:
> So MMR leafs + proofs are definitely essential for the client.

Yes, anything that you can prove about the latest state will require some sort of MMR proof. We don't necessarily need headers (since we might have some data directly in the MMR leaf), but yeah, generating a MMR proof is necessary for anything built "on top" of BEEFY.

We could bring some extra data to the commitment itself if necessary (maybe block hash along the block number?) - I guess this could simplify some use cases?

Reply:
Yeah, adding block hash to the commitment imo would make client development easier. But perhaps it will just be better (at least in terms of maintaining BEEFY) for all of us to use the same schema for all BEEFY clients on all chains. I can imagine that submitting header with public keys of `1024` Kusama BEEFY validators won't work for eth client (meaning tx size will be larger than 33kb).

So I'm leaning towards using MMR proofs, but I'd like to explore the option of generating these proofs outside of runtime to avoid relying on archive nodes. Of course if you or @acatangiu are insist on header-only light client, then we can alter commitment instead,

Reply:
> but I'd like to explore the option of generating these proofs outside of runtime to avoid relying on archive nodes.

This was easier than I've expected. Actually we don't need to do it outside of runtime - we only need to avoid touching any MMR-relevant storage items. I've implemented it here: https://github.com/paritytech/substrate/tree/sv-beefy-generate-historical-proof. Going to open PR on monday.

Reply:
upd: https://github.com/paritytech/substrate/pull/11230 and https://github.com/paritytech/polkadot/pull/5334 - don't have a time to work on that now, so I'm closing